File: editors/intellij/src/main/kotlin/dev/foxxycode/intellij/ui/ProjectRelativePaths.kt
package dev.foxxycode.intellij.ui

import java.io.File
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Turns files carried by a drag-and-drop onto the FoxxyCode panel into short,
 * project-relative POSIX paths for the composer `@`-mention.
 *
 * Kept free of IntelliJ APIs so it is unit-testable without the platform
 * (see ProjectRelativePathsTest); [FoxxyCodeBrowserPanel] owns the DnD plumbing.
 */
object ProjectRelativePaths {

    /**
     * Project-relative POSIX paths for [files], in input order and de-duplicated.
     * Directories, paths outside [basePath], and anything that cannot be
     * relativized are dropped, so an unrelated drag simply yields nothing.
     */
    fun relativize(basePath: String?, files: List<File>): List<String> {
        val base = basePath?.trim().orEmpty()
        if (base.isEmpty() || files.isEmpty()) return emptyList()
        val root: Path = try {
            Paths.get(base)
        } catch (e: Exception) {
            return emptyList()
        }
        val out = LinkedHashSet<String>()
        for (f in files) {
            if (f.isDirectory) continue
            val rel = try {
                root.relativize(f.toPath()).toString().replace('\\', '/')
            } catch (e: Exception) {
                continue
            }
            if (rel.isEmpty() || rel.startsWith("..")) continue
            out.add(rel)
        }
        return out.toList()
    }
}

File: editors/vscode/src/ide/projectRelativePaths.ts
/** Pure path helpers behind "Add to FoxxyCode". Kept free of any `vscode`
 *  import so they can be unit-tested in a plain Node environment; mirrors
 *  `editors/intellij/.../ui/ProjectRelativePaths.kt`. */

import * as nodePath from "path";
import * as fs from "fs";

export interface RelativizeOptions {
  /** Directory probe; the default stats the path and treats errors as "not a directory". */
  isDirectory?: (absPath: string) => boolean;
  /** Path implementation (tests inject `path.win32` / `path.posix`); defaults to the host's. */
  pathImpl?: nodePath.PlatformPath;
}

function defaultIsDirectory(absPath: string): boolean {
  try {
    return fs.statSync(absPath).isDirectory();
  } catch {
    return false;
  }
}

/** Workspace-relative POSIX paths for `absPaths`, in input order and de-duplicated.
 *  Directories, the root itself, paths outside `basePath`, and anything that
 *  cannot be relativized are dropped, so an unrelated selection simply yields
 *  nothing. A blank `basePath` yields nothing as well. */
export function relativizeWorkspacePaths(
  basePath: string | undefined | null,
  absPaths: readonly string[],
  opts: RelativizeOptions = {},
): string[] {
  const base = (basePath ?? "").trim();
  if (base === "" || absPaths.length === 0) return [];
  const p = opts.pathImpl ?? nodePath;
  const isDirectory = opts.isDirectory ?? defaultIsDirectory;
  const out = new Set<string>();
  for (const abs of absPaths) {
    const target = (abs ?? "").trim();
    if (target === "" || isDirectory(target)) continue;
    let rel: string;
    try {
      rel = p.relative(base, target);
    } catch {
      continue;
    }
    // "" is the root itself; ".." leads outside; an absolute result is a
    // different Windows drive.
    if (rel === "" || rel === ".." || rel.startsWith(".." + p.sep) || p.isAbsolute(rel)) continue;
    out.add(rel.split(p.sep).join("/"));
  }
  return [...out];
}
